packages/story-editor/src/app/story/useStory.js

```diff
@@ -27,4 +27,19 @@ function useStory(selector) {
   return useContextSelector(Context, selector ?? identity);
 }
 
+const DELIMITER = ':::';
+/**
+ * Returns an array of current page's story element ids. Only rerenders when there are
+ * element reorders, deletions or additions
+ *
+ * @return {Array} array of current pages story element ids in order
+ */
+export function useCurrentPageElementIds() {
+  const elementIds = useStory((v) => {
+    const elements = v.state.currentPage?.elements || [];
+    return elements.map((element) => element.id).join(DELIMITER);
```

Comment:
What‘s the secret sauce behind joining and splitting by this delimiter?
Can this be made simpler?
If not, can it be explained in a comment?

Reply:
Yeah, good call, I can absolutely explain it in a comment.

The idea behind this is the same rationale as the rest of the changes here which is that `useContextSelector` limits re-renders by doing strict equality checks amongst return values of the selector. If that equality check is false, it forces a re-render with the updated state:
<img width="852" alt="Screen Shot 2021-12-24 at 11 04 06 AM" src="https://user-images.githubusercontent.com/35983235/147366132-feff6e3b-c90f-451d-b361-48c8ac20590b.png">

src (don't have to focus on it too much, but can see through the comments that it's basically forcing a re-render with the updated selected value if referential equality changes):
https://github.com/dai-shi/use-context-selector/blob/7ca35cf68d08ba930a8b91c129ab00d1798300b9/src/index.ts#L144-L180

I tried to explain it in the doc string, but can put more technical detail comments in the implementation.

Basically if a component doesn't need element properties for its implementation, but does need something like the number of elements and their respective identifiers, this setup prevents that component from rendering when it doesn't need to.

ie
- there are 3 elements [{ id: 'a', ... }, { id: 'b', ... }, {id: 'c', ... }]
- a component can use this and get the info that the elements follow this ['a', 'b', 'c']
- we update element `b` to have a new `backgroundColor` property
- the component we're using this in would otherwise have an empty re-render that costs some time and would trigger a possible empty re-render in all of it's descendants. (could be less than 1ms per element, but if our component has a subtree of 100 elements, that adds up)
- with this hook use context selector would check the string `a:::b:::c` to the string `a:::b:::c` and not trigger a re-render in the component we're using this in.

**tldr;** it's a matter of limiting renders to only elements that will equate to a visual update on the screen.

Reply:
> The idea behind this is the same rationale as the rest of the changes here which is that useContextSelector limits re-renders by doing strict equality checks amongst return values of the selector. If that equality check is false, it forces a re-render with the updated state:

An alternative could also be using custom equality checks (here and elsewhere where applicable), especially if `useCurrentPageElementIds` is used in one place only, perhaps would make it more simple and avoid adding helpers just for getting past the default equality check. Thoughts?

Reply:
Got it, so just returning a string to make the equality check work. Just happening to use a custom delimiter to make it more robust.

Apologies, I was on mobile when I wrote this comment, but now with the additional context when seeing the whole code + the explanation things are more clear.

In that case, I'd have suggested the same thing as Miina, using a custom equality function, which can be passed as the second argument to `useStory`.

Also, if this used more often, we could consider storing the current page element list in the StoryProvider itself so we could do something like `useStory((value) => value.state.currentPageElementIds)`.

Either method would help avoiding adding helpers like that.

Reply:
@littlemilkstudio We are currently using a custom equality function in one place, by just accepting it in our hook wrapper:
https://github.com/google/web-stories-wp/blob/2d0a090015349a786e676864c5652df5a134893a/packages/story-editor/src/components/dropTargets/useDropTargets.js#L26-L28

And then adding the custom equality function here (also for performance reasons):
https://github.com/google/web-stories-wp/blob/2d0a090015349a786e676864c5652df5a134893a/packages/story-editor/src/components/library/panes/media/common/innerElement.js#L123-L134

Perhaps we could just accept the equality function for [`useStory`](https://github.com/google/web-stories-wp/blob/main/packages/story-editor/src/app/story/useStory.js#L26) as well?

Reply:
@miina thanks for the reference. didn't realize we added a wrapper around `useContextSelector` to enable other equality functions.

Since shallow equals is the default comparison method, just doing something like:
```
const currentPageElementIds = useStory(({state}) => state.currentPage.elements.map(el => el.id))
```
should achieve the same result we're getting here with joining and splitting with a delimiter. No need for this custom hook with a delimiter. Thanks for walking through this with me

Reply:
Ahh never mind. the version of `shallowEqual` we're using only applies to objects, it's not actually shallow equaling arrays:

https://github.com/gaearon/react-pure-render/blob/master/src/shallowEqual.js

Reply:
guess I do need to write a different equality function that basically does shallow equal for arrays and use it here

Reply:
updated our shallow-equal to work with arrays as well using this package:
https://www.npmjs.com/package/shallow-equal